base/strings/string.jl

```diff
@@ -190,13 +190,47 @@ end
 end
 
 ## checking UTF-8 & ACSII validity ##
+# This table is used by the shift base DFA validation for UTF-8
+const dfa_table = [      
+       fill(UInt64(384), 128);
+       fill(UInt64(6860978528477184), 16);
+       fill(UInt64(107228354863104), 16);
+       fill(UInt64(107202588205056), 32);
+       fill(UInt64(0), 2);
+       fill(UInt64(768), 30);
+       fill(UInt64(1152), 1);
+       fill(UInt64(1536), 12);
+       fill(UInt64(1920), 1);
+       fill(UInt64(1536), 2);
+       fill(UInt64(2304), 1);
+       fill(UInt64(2688), 3);
+       fill(UInt64(3072), 1);
+       fill(UInt64(0), 11)
+]::Vector{UInt64}
+
+# This is a shift based utf-8 DFA
+function _isvalid_utf8(bytes::Vector{UInt8})
+    f(byte) = @inbounds dfa_table[byte]
+    op(state, byte_dfa) = byte_dfa >> (state & UInt64(63))
+    return mapfoldl(f, op, bytes, init=UInt64(6)) == UInt64(6)
+end
+
+_isvalid_utf8(s::Union{String,Vector{UInt8},FastContiguousSubArray{UInt8,1,Vector{UInt8}}}) = _isvalid_utf8(unsafe_wrap(Vector{UInt8}, s))
 
-byte_string_classify(s::Union{String,Vector{UInt8},FastContiguousSubArray{UInt8,1,Vector{UInt8}}}) =
-    ccall(:u8_isvalid, Int32, (Ptr{UInt8}, Int), s, sizeof(s))
+# Classifcations of string
     # 0: neither valid ASCII nor UTF-8
     # 1: valid ASCII
     # 2: valid UTF-8
+function byte_string_classify(s::Union{String,Vector{UInt8},FastContiguousSubArray{UInt8,1,Vector{UInt8}}}; ascii_fasttrack = true )
+    bytes = unsafe_wrap(Vector{UInt8}, s)
+    ascii_fasttrack && all(c -> iszero(c & 0x80), bytes) && return 1
```

Comment:
can we make this track the first non ascii character so we don't have to rescan these characters for the valid utf8 check?

Reply:
I don't think this should be an issue it will just ad another subarray

Reply:
I am almost inclined to remove the fastpath, it was something that was in the c version in base and I included it because of that.    I think the mainpath is fast enough.

I have a working version that remembers the first non ascii location but the findfirst operation is so slow that is a hindrance.  Also this seems to be one of those functions that just happens to be very sensitive to the optimizations that get applied because below are two examples with the better fast path that give very different answers.
```julia
function byte_string_classify(bytes::Vector{UInt8}; ascii_fasttrack = true )
    if ascii_fasttrack
        first_non_ascii = findfirst(c -> c & 0x80 == 0x80, bytes) 
        first_non_ascii == nothing && return 1
        valid = _isvalid_utf8(@view bytes[first_non_ascii:end])
    else
        valid = _isvalid_utf8(bytes)
    end
   
    return ifelse(valid, 2, 0)
end

julia> benchmark_utfvalid(isvalid);
UTF-8: isvalid2 1.4856232014088069 billion bytes/second

julia> benchmark_asciivalid(isvalid);
ASCII: isvalid2 1.5308402664867142 billion bytes/second
```

While 
```julia 

function byte_string_classify(bytes::Vector{UInt8}; ascii_fasttrack = true )
    if ascii_fasttrack
        # first_non_ascii = findfirst(c -> c & 0x80 == 0x80, bytes) 
        first_non_ascii = findfirst(c -> !iszero(c & 0x80), bytes) 
        first_non_ascii == nothing && return 1
        valid = _isvalid_utf8(@view bytes[first_non_ascii:end])
    else
        valid = _isvalid_utf8(bytes)
    end
    return ifelse(valid, 2, 0)
end

julia> benchmark_utfvalid(isvalid);
UTF-8: isvalid2 2.1941641887672376 billion bytes/second

julia> benchmark_asciivalid(isvalid);
ASCII: isvalid2 1.5262564959843006 billion bytes/second
```

Reply:
also it might be worth benchmarking long strings as well (where vectorization and memory bandwidth are more imporant)

Reply:
The incredible thing is that the operations inde the loop of the DFA is already reduced to two operation.
```julia
LBB0_3:                                 ## %L45
                                        ## =>This Inner Loop Header: Depth=1
; │││││││┌ @ essentials.jl:13 within `getindex`
	movzbl	(%rcx,%rsi), %edi
; ││││││└└
; ││││││ @ reduce.jl:62 within `_foldl_impl`
; ││││││┌ @ reduce.jl:100 within `MappingRF`
; │││││││┌ @ reduce.jl:86 within `BottomRF`
; ││││││││┌ @ /Users/.../dev/julia/julia/base/strings/string.jl:222 within `op`
; │││││││││┌ @ int.jl:502 within `>>`
	shrxq	%rbx, (%rdx,%rdi,8), %rbx
; ││││││└└└└
; ││││││ @ reduce.jl:60 within `_foldl_impl`
; ││││││┌ @ array.jl:893 within `iterate`
; │││││││┌ @ int.jl:494 within `<` @ int.jl:487
	incq	%rsi
	cmpq	%rsi, %rax
; │││││││└
	jne	LBB0_3

```

Reply:
I ended up removing the fast track option from byte_string_check but leaving the check.  Even at 10 MB it adds almost no overhead